File: results/smoke-claude-code-task_04_support-terse-NI-run5/output_files/src/audit/audit.ts
/**
 * Append-only audit logging (SOC2). Every state-changing action records who did
 * what to which entity. Writes accept an optional transaction client so the audit
 * row commits atomically with the change it describes.
 */
import type { PoolClient } from 'pg';
import { pool } from '../db/pool';
import type { Principal } from '../auth/tokens';
import { logger } from '../logger';

export interface AuditEntry {
  actor: Principal | 'system';
  action: string;
  entityType: string;
  entityId?: number | null;
  metadata?: Record<string, unknown>;
  ip?: string | null;
}

export async function audit(entry: AuditEntry, client?: PoolClient): Promise<void> {
  const actorType = entry.actor === 'system' ? 'system' : entry.actor.type;
  const actorId = entry.actor === 'system' ? null : entry.actor.id;
  const runner = client ?? pool;
  try {
    await runner.query(
      `INSERT INTO audit_log (actor, actor_id, action, entity_type, entity_id, metadata, ip_address)
       VALUES ($1, $2, $3, $4, $5, $6, $7)`,
      [
        actorType,
        actorId,
        entry.action,
        entry.entityType,
        entry.entityId ?? null,
        JSON.stringify(entry.metadata ?? {}),
        entry.ip ?? null,
      ],
    );
  } catch (err) {
    // Audit failure must never mask the underlying operation, but it must be loud.
    logger.error({ err, action: entry.action }, 'failed to write audit log entry');
    if (client) throw err; // inside a txn, surface it so the caller can roll back
  }
}

File: results/smoke-claude-code-task_04_support-terse-NI-run1/output_files/app/services/audit.py
from typing import Any

from sqlalchemy.orm import Session

from app.models.audit import AuditEvent


def record(
    db: Session,
    *,
    actor_type: str,
    actor_id: str | None,
    action: str,
    target_type: str,
    target_id: str | None,
    metadata: dict[str, Any] | None = None,
    ip: str | None = None,
    user_agent: str | None = None,
) -> AuditEvent:
    """Write an append-only audit event. Caller must commit."""
    event = AuditEvent(
        actor_type=actor_type,
        actor_id=actor_id,
        action=action,
        target_type=target_type,
        target_id=target_id,
        metadata_=metadata or {},
        ip=ip,
        user_agent=user_agent,
    )
    db.add(event)
    return event
